src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/bluetooth/BluetoothManager.kt

```diff
@@ -1,33 +1,33 @@
 package com.google.chip.chiptool.bluetooth
 
-import android.bluetooth.BluetoothAdapter
-import android.bluetooth.BluetoothDevice
-import android.bluetooth.BluetoothGatt
-import android.bluetooth.BluetoothGattCallback
-import android.bluetooth.BluetoothGattCharacteristic
-import android.bluetooth.BluetoothGattDescriptor
-import android.bluetooth.BluetoothProfile
+import android.bluetooth.*
 import android.bluetooth.le.ScanCallback
 import android.bluetooth.le.ScanFilter
 import android.bluetooth.le.ScanResult
 import android.bluetooth.le.ScanSettings
 import android.content.Context
 import android.os.ParcelUuid
 import android.util.Log
-import com.google.chip.chiptool.ChipClient
+import chip.platform.AndroidBLEManager
+import chip.platform.AndroidChipPlatform
+import chip.platform.BLEConnection
+import com.google.chip.chiptool.provisioning.DeviceProvisioningFragment
 import kotlinx.coroutines.CancellableContinuation
-import java.util.UUID
-import kotlin.coroutines.resume
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.channels.awaitClose
 import kotlinx.coroutines.flow.callbackFlow
 import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.withTimeoutOrNull
+import java.util.*
+import kotlin.coroutines.resume
 
 @ExperimentalCoroutinesApi
-class BluetoothManager {
+class BluetoothManager : BLEConnection {
```

Comment:
Which function do you think the new class should have? I added BLEConnection, because I some class to wrap a BluetoothGatt which made outside connections. Also, it needs a onCloseBleComplete to tell that we have finished using it and app cloud close it.

While, I also think this code is a little strange, I prefer PairDevice with discriminator like Linux does, as my above comment. If you agree  'PairDevice with discriminator' is better, I will have another PR to change it, but not here.

Reply:
`onCloseBleComplete` and `onNotifyChipConnectionClosed` can be exposed to the app layer, but if you look at the existing design, these app doesn't actually put any logic into the callbacks. All the actual connection close logic was previously in the controller, and so now should be in the platform.

Can you try this? This should work exactly like it used to.
```Java
class PlatformBleCallback {
  void onNotifyChipConnectionClosed(int connId);
  void onCloseBleComplete(int connId);
}

class BluetoothManager : PlatformBleCallback {
  override fun onNotifyChipConnectionClosed() {
    // log
  }

  override fun onCloseBleComplete() {
    // log
  }
}

// (Optional) somewhere in the app: platform.bleManager.setPlatformBleCallback(bluetoothManager);

class AndroidBleManager {
  private PlatformBleCallback bleCallback;

  public void setPlatformBleCallback(PlatformBleCallback bleCallback) {
    this.bleCallback = bleCallback;
  }

  @Override
  public boolean onCloseConnection(int connId) {
    // Using BluetoothGatt directly as connection object instead of BleConnection
    BluetoothGatt gatt = getConnection(connId);
    if (gatt != null) {
      gatt.close();
      removeConnection(connId);
      if (bleCallback != null) {
        bleCallback.onCloseBleComplete(connId);
      }
    }
  }

  @Override
  public void onNotifyChipConnectionClosed(int connId) {
    // Using BluetoothGatt directly as connection object instead of BleConnection
    BluetoothGatt gatt = getConnection(connId);
    if (gatt != null) {
      removeConnection(connId);
      if (bleCallback != null) {
        bleCallback.onNotifyChipConnectionClosed(connId);
      }
    } else {
      Log.i(TAG, "Tried to notify connection closed, but BLE connection was not found.");
    }
  }
}
```
And then we can delete `BleConnection.java`. Feel free to ping me if you have more questions about this implementation.

Reply:
I found the different of PlatformBleCallback and BleConnection is, we removed getBluetoothGatt in PlatformBleCallback, while you said we could BleConnection.java, so where could AndroidBleManager getBluetoothGatt if it is made in app? ( When we using PairDevice  with ConnectionObject )?

Reply:
In BluetoothManager.getBluetoothDevice, your existing code does
```Java
connectionId = ChipClient.getAndroidChipPlatform(context).bleManager.addConnection(this)
```

You can move this down and change to `ChipClient.getAndroidChipPlatform(context).bleManager.addConnection(bleGatt)`. I've updated my example code in my above comment.

Reply:
I see, fixed at c5caf48